src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs

```diff
@@ -16,23 +18,366 @@ public void Dispose()
+
+                case Interop.libcrypto.X509VerifyStatusCode.X509_V_ERR_INVALID_CA:
+                case Interop.libcrypto.X509VerifyStatusCode.X509_V_ERR_INVALID_NON_CA:
+                case Interop.libcrypto.X509VerifyStatusCode.X509_V_ERR_PATH_LENGTH_EXCEEDED:
+                case Interop.libcrypto.X509VerifyStatusCode.X509_V_ERR_KEYUSAGE_NO_CERTSIGN:
+                case Interop.libcrypto.X509VerifyStatusCode.X509_V_ERR_KEYUSAGE_NO_DIGITAL_SIGNATURE:
+                    return X509ChainStatusFlags.InvalidBasicConstraints;
+
+                case Interop.libcrypto.X509VerifyStatusCode.X509_V_ERR_INVALID_POLICY_EXTENSION:
+                case Interop.libcrypto.X509VerifyStatusCode.X509_V_ERR_NO_EXPLICIT_POLICY:
+                    return X509ChainStatusFlags.InvalidPolicyConstraints;
+
+                case Interop.libcrypto.X509VerifyStatusCode.X509_V_ERR_CERT_REJECTED:
+                    return X509ChainStatusFlags.ExplicitDistrust;
+
+                case Interop.libcrypto.X509VerifyStatusCode.X509_V_ERR_UNHANDLED_CRITICAL_EXTENSION:
+                    return X509ChainStatusFlags.HasNotSupportedCriticalExtension;
+
+                case Interop.libcrypto.X509VerifyStatusCode.X509_V_ERR_CERT_CHAIN_TOO_LONG:
+                case Interop.libcrypto.X509VerifyStatusCode.X509_V_ERR_OUT_OF_MEM:
+                    throw new CryptographicException();
+
+                default:
+                    Debug.Assert(false, "Unrecognized X509VerifyStatusCode:" + code);
+                    throw new CryptographicException();
+            }
+        }
+
+        internal static X509Certificate2Collection FindCandidates(
+            X509Certificate2 leaf,
+            X509Certificate2Collection extraStore)
+        {
+            X509Certificate2Collection candidates = new X509Certificate2Collection();
+            Queue<X509Certificate2> toProcess = new Queue<X509Certificate2>();
+            toProcess.Enqueue(leaf);
+
+            using (var rootStore = new X509Store(StoreName.Root, StoreLocation.LocalMachine))
+            using (var intermediateStore = new X509Store(StoreName.CertificateAuthority, StoreLocation.LocalMachine))
+            {
+                rootStore.Open(OpenFlags.ReadOnly);
+                intermediateStore.Open(OpenFlags.ReadOnly);
+
+                X509Certificate2Collection rootCerts = rootStore.Certificates;
+                X509Certificate2Collection intermediateCerts = intermediateStore.Certificates;
+
+                while (toProcess.Count > 0)
+                {
+                    X509Certificate2 current = toProcess.Dequeue();
+
+                    if (!candidates.Contains(current))
+                    {
+                        candidates.Add(current);
+                    }
+
+                    X509Certificate2Collection results = FindIssuer(
+                        current,
+                        extraStore,
+                        intermediateCerts,
```

Comment:
FindIssuer is defined to take a params array, so each call in this loop is going to allocate a new three-element array.  I'd suggest either changing the FindIssuer signature, or at least caching the array into a local so that it's allocated once rather than once per iteration.